main/src/main/java/org/mobilitydata/gtfsvalidator/validator/DuplicateFareRuleZoneIdFieldsValidator.java

```diff
@@ -41,27 +42,36 @@
 
   @Override
   public void validate(NoticeContainer noticeContainer) {
-    final Map<String, GtfsFareRule> fareRuleByZoneIdFieldsCombination =
+    final Map<Integer, GtfsFareRule> fareRuleByZoneIdFieldsCombination =
         new HashMap<>(fareRuleTable.entityCount());
     fareRuleTable
         .getEntities()
         .forEach(
             fareRule -> {
-              String fieldsCombination =
-                  fareRule.routeId()
-                      + fareRule.originId()
-                      + fareRule.containsId()
-                      + fareRule.destinationId();
-              if (fareRuleByZoneIdFieldsCombination.containsKey(fieldsCombination)) {
-                noticeContainer.addValidationNotice(
-                    new DuplicateFareRuleZoneIdFieldsNotice(
-                        fareRule.csvRowNumber(),
-                        fareRule.fareId(),
-                        fareRuleByZoneIdFieldsCombination.get(fieldsCombination).csvRowNumber(),
-                        fareRuleByZoneIdFieldsCombination.get(fieldsCombination).fareId()));
-              } else {
-                fareRuleByZoneIdFieldsCombination.put(fieldsCombination, fareRule);
+              int hash = getHash(fareRule);
+              GtfsFareRule otherFareRule = fareRuleByZoneIdFieldsCombination.get(hash);
+              if (otherFareRule == null) {
+                fareRuleByZoneIdFieldsCombination.put(hash, fareRule);
+                return;
               }
+              noticeContainer.addValidationNotice(
+                  new DuplicateFareRuleZoneIdFieldsNotice(
+                      fareRule.csvRowNumber(),
+                      fareRule.fareId(),
+                      otherFareRule.csvRowNumber(),
+                      otherFareRule.fareId()));
             });
   }
+
+  /**
+   * Returns the hashcode associated to the combination of `fare_rules.origin_id`,
```

Comment:
Backticks do not work in Javadoc, unfortunately. You can use `{@code ...}` or double quotes.